packages/kbn-unified-field-list/src/hooks/use_grouped_fields.ts

```diff
@@ -120,13 +126,30 @@ export function useGroupedFields<T extends FieldListItem = DataViewField>({
     };
 
     const selectedFields = sortedSelectedFields || [];
-    const sortedFields = [...(allFields || [])].sort(sortFields);
+
+    // Taking care of new fields that were ingested after the selected data view was loaded
+    // Those replace existing fields if updated, or are added to the list
+    const newFields = dataViewId
+      ? fieldsExistenceReader.getNewFields(dataViewId).filter(isCompatibleField || (() => true))
+      : [];
+    // Filtering out fields that e.g. Discover provides by analyzing the resultset, that are not part of the loaded DataView
+    // These can be replaced by the new fields, which are mapped correctly, and therefore can be used in the right way
+    const allFieldsExlNew = allFields
+      ? allFields.filter((field) => !newFields.some((newField) => newField.name === field.name))
+      : [];
+
+    const allFieldsInclNew = [...allFieldsExlNew, ...newFields] as unknown as T[];
```

Comment:
`allFields` in Lens is an array of `IndexPatternField` instead of `DataViewField`. So if we force here and combine both types, other functionality might be broken for the new items in the list as certain properties would be missing in field objects.

I would suggest to exclude Lens from the scope of this PR and apply the logic only if `allFields` is an array of `DataViewField` type items.

Reply:
I'm not 100% finished with an alternative approach, it's already green. So I refactored for consumers to pass down a new method `getNewFieldsBySpec` to `useGroupedFields`. This fixes the typing in this case and also allows Lens to get new fields. This refactoring is part of https://github.com/elastic/kibana/pull/172329/files/12616b4fdda3b84254a100df23c19a32b51f42ee . I'm interested in your opinion

Reply:
Yes, this could work 👍